### PR TITLE
[Experiment] Rework child classes detection on DynamicSourceLocatorProvider

### LINKED ISSUE #8557: Parent class return type is added which is breaking the code
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 1.0.3            |

From the example below, rector is adding return type to the parent method.
image

As child can have different return type it is causing PHP error

## Minimal PHP Code Causing Issue

```php
<?php

abstract class ParentDemoFile
{
    protected function formatAmount(float $amount)
    {
        return $amount;
    }
}

 class DemoFile extends ParentDemoFile
{
    protected function formatAmount(float $amount): int
    {
        return (int) (\round($amount, 2) * 100);
    }
}
```

rector.php
```php
<?php

declare(strict_types=1);

use Rector\Config\RectorConfig;
use Rector\Set\ValueObject\LevelSetList;

return RectorConfig::configure()
    ->withSets([
        LevelSetList::UP_TO_PHP_74,
    ])
    ->withSkip([
        // ALL of these rules doing the same breaking change
        //\Rector\TypeDeclaration\Rector\ClassMethod\NumericReturnTypeFromStrictScalarReturnsRector::class,
        //\Rector\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromStrictScalarReturnExprRector::class,
        //\Rector\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromStrictParamRector::class,
    ])
    ->withPreparedSets(
        true,
        true,
        true,
        true,
        true,
        true,
        true,
        true,
        true,
    );
```

## Expected Behaviour

Do not break PHP code

### PR DESCRIPTION
**Pull request by samsonasik**
@nuryagdym  @TomasVotruba  this is experimental for https://github.com/rectorphp/rector/issues/8557

Closes https://github.com/rectorphp/rector/issues/8557

### CODE REVIEW

File: src/NodeTypeResolver/Reflection/BetterReflection/SourceLocatorProvider/DynamicSourceLocatorProvider.php [hunk 1]
```diff
@@ -76,6 +79,21 @@ public function provide(): SourceLocator
...
+            try {
+                $sourceLocator->locateIdentifiersByType($reflector, new class() extends IdentifierType {
+                    public function isClass(): bool
+                    {
+                        return true;
+                    }
+                });
+            } catch (CouldNotReadFileException) {
+            }
+        }
```
Comment (samsonasik) [inline src/NodeTypeResolver/Reflection/BetterReflection/SourceLocatorProvider/DynamicSourceLocatorProvider.php]: @TomasVotruba this is what I mean, previously, on very first run, the `classes` seems overlapped by which one file that scanned first, then on 2nd run, the class child will be detected.

This ensure `Reflector->reflectClass()` called on all `ClassLike` so it exists in the `classes` property in `MemoizingReflectionProvider`

I will try add back `getChildrenOfClassReflection()` method to `FamilyRelationsAnalyzer`, add fixture test, then let @nuryagdym manually try the patch.

**Comment by samsonasik** [comment]
Ok, here investigation:

**1** Tried back to Rector `0.19.15` before the `FinalizeClassesWithoutChildrenRector` is deprecated, tested with:


```
git clone git@github.com:samsonasik/ci4-album.git rector-0.x
```
 got output:

```diff
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector --dry-run
 49/49 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
2 files with changes
====================

1) src/Domain/Exception/DuplicatedRecordException.php:12

    ---------- begin diff ----------
@@ @@

 use DomainException;

-class DuplicatedRecordException extends DomainException
+final class DuplicatedRecordException extends DomainException
```

which shouldnot as the class has child. I will check if there is possible way to "init" the classes to be resolved by phpstan Reflection.

**Comment by samsonasik** [comment]
@TomasVotruba @staabm It somehow can be resolved by patch:

```diff
-            if (!$classReflection->isSubclassOf($desiredClassReflection->getName())) {
+            if ($desiredClassReflection->getName() !== $classReflection->getName() && !$classReflection->isSubclassOf($desiredClassReflection->getName())) {
                continue;
            }
```

it seems `isSubclassOf()` definition make same class name as true on the removed `FamilyRelationsAnalyzer::getChildrenOfClassReflection()` on rector 0.19.5

https://github.com/rectorphp/rector-src/blob/171fddb1a1e9c54e194130ffe039809721d1f465/src/FamilyTree/Reflection/FamilyRelationsAnalyzer.php#L39

I will try apply the child detection back.

**Comment by staabm** [comment]
I cannot follow, but you are right that [`isSubclassOf`](https://github.com/phpstan/phpstan-src/blob/c983705c3abc9c7fdfebcd29fae389b5e20c2f2c/src/Reflection/ClassReflection.php#L804) requires a actual subclass and does not include equality.

you can use [`is`](https://github.com/phpstan/phpstan-src/blob/c983705c3abc9c7fdfebcd29fae389b5e20c2f2c/src/Reflection/ClassReflection.php#L799) if you need the equality check to also return `true`.

**Comment by samsonasik** [comment]
It seems the `classes` only get the paths that scanned:

```php
        $classReflections = $this->privatesAccessor->getPrivateProperty($this->reflectionProvider, 'classes');
        var_dump(count($classReflections));
```

Running on actual path only got 6 classes:

```bash
➜  ci4-album git:(rector-0.x) vendor/bin/rector process src/Domain/Exception/RecordNotFoundException.php  --dry-run --clear-cache
 0/1 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%int(6)
string(46) "Album\Domain\Exception\RecordNotFoundException"
string(15) "DomainException"
string(14) "LogicException"
string(9) "Exception"
string(9) "Throwable"
string(10) "Stringable"
```

which should not, I will continue investigating if there is a way to init somewhere.

**Comment by staabm** [comment]
ClassReflection only works on paths in the analyzed path or scan*  configurations, thats expected and correct from PHPStan point of view (and composer autoloading I guess).

**Comment by samsonasik** [comment]
@staabm thank you, I remember that on old phpstan, the code is working fine without the tweak, and on some point of phpstan release, it no longer working, I will continue investigating as it is interesting, I expect it should work with at least run on no path provided:

```
vendor/bin/rector
```

which means all classes loaded.

**Comment by samsonasik** [comment]
@TomasVotruba @staabm I finally got it working prototype 🎉 , tested on Rector 0.19.5 before the `FinalizeClassesWithoutChildrenRector` deprecated that test sub classes.

the `getClass()` need to be called to trigger collect `classes` property of `MemoizingReflectionProvider` https://github.com/rectorphp/rector-src/pull/5735/commits/9a2fdbd90bd68d0b9c7a556efea4e8a86465afd9

**Before**

```diff
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector process --dry-run --clear-cache
 49/49 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
2 files with changes
====================

1) src/Domain/Exception/DuplicatedRecordException.php:12

    ---------- begin diff ----------
@@ @@

 use DomainException;

-class DuplicatedRecordException extends DomainException
+final class DuplicatedRecordException extends DomainException
```

**After**

```
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector process --dry-run --clear-cache
 49/49 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!                                                                                                   
```

 I will check segmentation fault on package tests

**Comment by samsonasik** [comment]
@nuryagdym @TomasVotruba @staabm it finally working again 🎉 🎉 🎉 , here the story:

-------

Long time ago, the children classes detection mostly always working, and mostly just require `composer dump-autoload`, until on some point release of phpstan, it no longer working, even autoload the classes early,

Turns out, the PHPStan has `NewOptimizedDirectorySourceLocator` which seems a brand new source locator, which somehow, make the PHPStan `ReflectionProvider` require to touch it to trigger `classes` to make it collected.

I tested in my own module: https://github.com/samsonasik/ci4-album/pull/16 and it seems manually apply the code, and it finally working fine.

This PR added back the feature, and the handle in `ClassMethodReturnTypeOverrideGuard` which detect if child classes override the return type, it will not be allowed, see fixture https://github.com/rectorphp/rector-src/pull/5735/files#diff-d52d2b21cf8f834596c1ab5e7bee4fae94181a109dec119b506d538f8017f8d6

and **e2e** test for the proof for `bin/rector` command usage https://github.com/rectorphp/rector-src/pull/5735/commits/f5f5973fe64c3031e661500038f11423e5de6c21

**Limitation:**

- Directly pass to single file will only get `classes` on single file, which expected.

```diff
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector process src/Domain/Exception/DuplicatedRecordException.php --dry-run --clear-cache
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
1 file with changes
===================

1) src/Domain/Exception/DuplicatedRecordException.php:12

    ---------- begin diff ----------
@@ @@

 use DomainException;

-class DuplicatedRecordException extends DomainException
+final class DuplicatedRecordException extends DomainException
```

so the run needs to apply on the parent and child classes contains, eg, both parent and child classes files:

```bash
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector process src/Domain/Exception/DuplicatedRecordException.php src/Domain/Track/TrackDuplicatedRectorException.php  --dry-run --clear-cache
 2/2 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!                                                                                                   
                                                                                                                        
```

or the directory containst both:

```bash
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector process src/ --dry-run --clear-cache 
 36/36 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!                                                                                                                                                                                                                    
```

or just without arg:

```bash
➜  ci4-album git:(rector-0.x) ✗ vendor/bin/rector
 36/36 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!     
```

If this PR merged, we can **un-deprecate** `FinalizeClassesWithoutChildrenRector` and make it works out off the box, in separate PR.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**